conda-store-server/conda_store_server/build.py

```diff
@@ -110,6 +111,85 @@ def build_environment(conda_command, environment_filename, conda_prefix):
     )
 
 
+def build_lock_environment(lock_filename, conda_prefix):
+    return subprocess.check_output(
+        ["conda-lock", "install", "--prefix", conda_prefix, lock_filename],
+        stderr=subprocess.STDOUT,
+        encoding="utf-8",
+    )
+
+
+def fetch_and_extract_packages(conda_store, lock_filename):
+    """Fetch links from a conda-locked build recipe and then
+    gets a filelock on the required folder.
+    """
+    import tarfile
+
+    prefix = pathlib.Path("/opt/conda/pkgs/")
+
+    try:
+        prefix.exists()
+    except FileNotFoundError as e:
+        conda_store.log.error(
+            f"The conda prefix {prefix} does not exist. Traceback: {e}"
+        )
+
+    spec = dict()
+
+    with open(pathlib.Path(lock_filename)) as f:
+        spec = yaml.safe_load(f)
+
+    packages_searched = 1
+
+    for p in spec["package"]:
+        url = p["url"]
+
+        filename = url.split("/")[-1:][0]
+        filepath = prefix.joinpath(filename)
+        lockfile = filelock.FileLock(f"{filepath.__str__()}.lock")
+
+        num_files_log = f"{packages_searched} of {len(spec['package'])}"
+
+        with lockfile:
+
+            if filepath.exists():
+                conda_store.log.info(f"SKIPPING {filename} | FILE EXISTS")
+                packages_searched += 1
+
+            else:
+                conda_store.log.info(f"DOWNLOAD {filename} | {num_files_log}")
+                res = requests.get(url, stream=True)
+
+                with open(filepath, "wb") as tarball:
+                    tarball.write(res.raw.read())
```

Comment:
This is inefficient since it will read the entire file into memory. Use https://docs.python.org/3/library/shutil.html#shutil.copyfileobj. Which will copy to disk in chunks efficiently.

Reply:
Done!